staging/src/k8s.io/client-go/tools/portforward/websocket-connection.go

```diff
@@ -0,0 +1,387 @@
+		// such as ping, pong, or close, this call is necessary for these
+		// message types to be processed. There can only be one reader
+		// at a time, so this reader loop must *not* be run concurrently;
+		// there is no lock for reading. Calling "NextReader()" before the
+		// current reader has been processed will close the current reader.
+		// If the heartbeat read deadline times out, this "NextReader()" will
+		// return an i/o error, and error handling will clean up.
+		messageType, r, err := c.conn.NextReader()
+		if err != nil {
+			websocketErr, ok := err.(*gwebsocket.CloseError)
+			if !ok || websocketErr.Code != gwebsocket.CloseNormalClosure {
+				c.closeAllStreamReaders(fmt.Errorf("next reader: %w", err))
+			}
+			return
+		}
+		// Throttle reading from the websocket connection (note: only one Reader
+		// on the connection can be active at a time).
+		if c.maxBytesPerSec > 0 {
+			r = flowrate.NewReader(r, c.maxBytesPerSec)
+		}
+		// All remote command protocols send/receive only binary data messages.
+		if messageType != gwebsocket.BinaryMessage {
+			c.closeAllStreamReaders(fmt.Errorf("unexpected message type: %d", messageType))
+			return
+		}
+		// Initially, read the websocket stream headers.
+		wsStreamHeaders, err := readWsStreamHeaders(r)
+		if err != nil {
+			c.closeAllStreamReaders(fmt.Errorf("read stream id: %w", err))
+			return
+		}
+		klog.V(5).Infof("websocket stream headers read: %s", wsStreamHeaders.String())
+		streamType := wsStreamHeaders.StreamType
+		streamID := wsStreamHeaders.StreamID
+		// StreamCreate signal means the other websocket connection endpoint
+		// has created a new stream.
+		if streamType == StreamCreate {
+			klog.V(4).Infof("stream create signal: %d", streamID)
+			// Create the stream, but do not send StreamCreate signal.
+			stream, err := c.createStream(wsStreamHeaders.Headers)
+			if err != nil {
+				c.closeAllStreamReaders(fmt.Errorf("creating websocket stream: %v", err))
+				return
+			}
+			// Queue newly created websocket stream onto channel.
+			klog.V(5).Infof("queueing stream on channel: %d", streamID)
+			streamCreateCh <- stream
+			continue
+		} else if streamType == StreamClose {
+			// Read the next byte, which is the stream id.
+			klog.V(5).Infof("stream close signal: %d", streamID)
+			s := c.getStream(streamID)
+			if s != nil {
+				s.writePipe.Close() //nolint:errcheck
+			} else {
+				klog.V(6).Infof("Unknown stream id during close %d--discarding message", streamID)
+			}
+			continue
+		}
+		klog.V(5).Infof("received stream %d", streamID)
+		s := c.getStream(streamID)
+		if s == nil {
+			klog.Errorf("Unknown stream id %d, discarding message", streamID)
+			continue
+		}
+		for {
+			klog.V(6).Infof("reading into buffer (%d)", streamID)
+			nr, errRead := r.Read(readBuffer)
+			if nr > 0 {
+				klog.V(6).Infof("writing into stream pipe (%d): %s", streamID, readBuffer[:nr])
+				_, errWrite := s.writePipe.Write(readBuffer[:nr])
```

Comment:
This will block if the corresponding http stream hasn't been read by the paired SPDY side goroutine?  I'm a little rusty on my SPDY, but is there any way the channel send on the SPDY side gets blocked by the kubelet (because a server on the other side of portforward proxy in the pod stops reading)?  If that happens, then this will block and the connection won't make progress reading other messages, which blocks all other concurrent requests, right?

It's been at least 3 years since I went through this last, so I might be too rusty, but:

The original reason SPDY was desirable for proxy was that it provided end to end backpressure - if kubelet couldn't write a buffer to a specific port, then SPDY would block all the way at the client end (at least, that's what I remember....).  That prevented any head of line blocking, because other channels would be able to make progress end to end with SPDY.

But in this implementation, if SPDY on kubelet puts backpressure on any given channel, it will do so on the apiserver, which will block the websocket, which will prevent concurrent traffic from passing through?  

The test case I think would be:

Start proxy over websocket, send two simultaneous requests to a server /a and /b (where /b is longer than the size of the receive side SDPY buffers), but have the server `sleep(long)` on /b.  /a might return, but any subsequent requests should block.

Reply:
> but any subsequent requests should block.

are you saying you think they *will* block with this impl, but we don't want them to, and the spdy version would not?

Reply:
(trying to bring all of this back into memory)

SPDY was chosen originally because it implements flow control for multiple clients, so concurrent requests over the same channel for proxy would be asynchronous with respect to each other (i.e. multiple tunnels).  SPDY implements [flow control](https://www.chromium.org/spdy/spdy-protocol/spdy-protocol-draft3-1/#the-flow-control-window) and prevents senders from sending larger than the destination buffers (kubelet, or in this case, apiserver, it's the same problem) which when filled cause other channels to block.

Because websockets is a "dumb" channel you end up having at least three options:

1. Don't allow multiple simultaneous connections over websockets (easiest)
2. Implement flow control over websockets - maintain a flow window that you reduce each time you write, and have the receiver send WINDOW_UPDATE messages / multiplex them on returning channel (optimization)
3. Treat the websocket as a dumb channel, and run SPDY over it.

Since one of the original goals for websocket endpoint was to make it easy to implement clients in browsers / other languages, 3 is somewhat challenging (are there any SPDY js libraries), as is 2 (you'd need to implement a custom flow control client in each language).

I think this is why we "gave up" last time - 1 felt too onerous.

Reply:
> are you saying you think they will block with this impl, but we don't want them to, and the spdy version would not?

If you have to manage fixed size buffers in the apiserver (connected to the websocket), and you have slow readers on the kubelet side (connected via SPDY), if a client then sends data to the apiserver and it makes it onto the websocket channel, the entire websocket is "blocked" until the buffer is emptied to the kubelet over SPDY.

1. Client 1 sends N bytes to server on port 1
2. Server reads N bytes off websocket to buffer
3. Server tries to send N bytes to kubelet
4. Kubelet accepts N bytes, SPDY window is reduced to 0 from apiserver
5. Kubelet tries to write N bytes to connection, connection is blocked
6. Client 1 sends N bytes to server
7. Server reads N bytes off websocket to buffer
8. Server tries to send N bytes to kubelet, but window is zero
9. Client 1 sends N bytes to server
10. Server reads N bytes off websocket, tries to write to buffer, buffer is full
11. Server is now blocked - it can't get rid of the buffer (without holding more memory), it's can't drop the bytes (no client retry protocol)
12. Client 2 sends N bytes to server on port 2, but is blocked until remote client reads N bytes on port 1 and unblockes server

Reply:
We could always document "multi-port websockets on apiserver will block if one client attempts to send more than XXX KB, so don't use multi-port websockets for anything latency sensitive" as an option 4

Reply:
> This will block if the corresponding http stream hasn't been read by the paired SPDY side goroutine? I'm a little rusty on my SPDY, but is there any way the channel send on the SPDY side gets blocked by the kubelet (because a server on the other side of portforward proxy in the pod stops reading)? If that happens, then this will block and the connection won't make progress reading other messages, which blocks all other concurrent requests, right?
> 

There is no websocket connection blocking--up to a point. The goroutine that reads the pipe/stream is spawned on the apiserver for each request here:

https://github.com/kubernetes/kubernetes/blob/02c611130007c581d4b68760942905ecd3f5650a/staging/src/k8s.io/apiserver/pkg/util/proxy/portforward/streamtranslator.go#L384-L396

This `io.Copy()` should not block, even if the upstream SPDY stream is blocked. The `io.Copy()` buffer is the same size as the websocket connection read buffer (32 KB), so the *initial* websocket stream `Read()` will never block--the websocket connection will continue in its loop. Very large requests may be problematic. But we can make this `io.Copy()` buffer even bigger if we need to using `io.CopyBuffer()`.

The cost of this implementation is that we've moved the problem from blocking to almost unbounded concurrency. We basically spawn the same number of goroutines in the apiserver that we have on all the clients. Initially, I implemented a concurrent request throttling mechanism using a condition variable on the data structure which stores each requests streams. A simple script sending 100k requests validated this request throttling mechanism. For alpha, however, I decided to remove this request throttling because it wasn't clear it was necessary, and it would need significant benchmarks to get the concurrency limit correct. We can revisit this decision if needed.

> It's been at least 3 years since I went through this last, so I might be too rusty, but:
> 
> The original reason SPDY was desirable for proxy was that it provided end to end backpressure - if kubelet couldn't write a buffer to a specific port, then SPDY would block all the way at the client end (at least, that's what I remember....). That prevented any head of line blocking, because other channels would be able to make progress end to end with SPDY.
> 
> But in this implementation, if SPDY on kubelet puts backpressure on any given channel, it will do so on the apiserver, which will block the websocket, which will prevent other concurrent traffic from passing through?
> 
> The test case I think would be:
> 
> Start proxy over websocket, send two simultaneous requests to a server /a and /b (where /b is longer than the size of the receive side SDPY buffers), but have the server `sleep(long)` on /b. /a might return, but any subsequent requests should block.

I have put together a simple test which I believe addresses this issue.

1. I created a simple server which has a `/blocking` endpoint which sleeps for 60 seconds before returning `blocked file` text. Other endpoints of the server return basic html files (including one large file > 620KB). I deployed this server to my local cluster, and the server listens on port 3333.
2. `kubectl port-forward test-server 3000:3333 4000:3333` starting the port-forward websocket and SPDY connection.
```$ ./kubectl port-forward -v=4 test-server 3000:3333 4000:3333
I0216 05:58:52.322913 3848705 portforward.go:143] PortForward WebSockets Enabled
I0216 05:58:52.323136 3848705 websocket-dialer.go:77] Remote Port: 3333
I0216 05:58:52.323152 3848705 websocket-dialer.go:77] Remote Port: 3333
I0216 05:58:52.323162 3848705 websocket-dialer.go:83] Before WebSocket Upgrade Connection...
I0216 05:58:52.342215 3848705 websocket-dialer.go:92] negotiated protocol: v2.portforward.k8s.io
Forwarding from 127.0.0.1:3000 -> 3333
Forwarding from [::1]:3000 -> 3333
Forwarding from 127.0.0.1:4000 -> 3333
Forwarding from [::1]:4000 -> 3333
I0216 05:58:52.343012 3848705 portforward.go:309] Before listener.Accept()...
I0216 05:58:52.343041 3848705 portforward.go:309] Before listener.Accept()...
I0216 05:58:52.343085 3848705 portforward.go:309] Before listener.Accept()...
I0216 05:58:52.343162 3848705 portforward.go:309] Before listener.Accept()...
```
4. Send a very large port-forward request, which is blocked
```
$ curl -X POST -d @post-64kb.dat http://localhost:3000/blocking
```
5. In another terminal, run 400 regular `curl` requests, which hit four different endpoints (using two different local ports).
```
$ ./portforward-test-2
Starting port forward test: 3000:3333 4000:3333

Temp Dir: /tmp/tmp.fSrTtNGsvo
Errors Dir: /tmp/tmp.YZt6CswDa6

.........10
.........20
.........30
.........40
.........50
.........60
.........70
.........80
.........90
.........100

Num Requests: 400
Num Errors: 0

Cleaning up...

Finished
```

This script did not block. The final output from the `kubectl port-forward ...` logging, looks like
```
Handling connection for 3000
I0216 06:04:38.525571 3848705 websocket-connection.go:130] CreateStream: 798
I0216 06:04:38.525742 3848705 websocket-connection.go:130] CreateStream: 799
I0216 06:04:38.528320 3848705 portforward.go:395] Closing DataStream...Writing Complete
I0216 06:04:38.528427 3848705 portforward.go:388] Datastream Reading Complete...Closing remoteDone channel
I0216 06:04:38.528448 3848705 portforward.go:420] handle FINISHED...
I0216 06:04:38.528499 3848705 websocket-connection.go:161] RemoveStream: 799
I0216 06:04:38.528511 3848705 websocket-connection.go:161] RemoveStream: 798
I0216 06:04:38.540823 3848705 portforward.go:318] Accepted; handling connection on {4000 3333}
I0216 06:04:38.540865 3848705 portforward.go:309] Before listener.Accept()...
Handling connection for 4000
I0216 06:04:38.540917 3848705 websocket-connection.go:130] CreateStream: 800
I0216 06:04:38.541070 3848705 websocket-connection.go:130] CreateStream: 801
I0216 06:04:38.547116 3848705 portforward.go:395] Closing DataStream...Writing Complete
I0216 06:04:38.547211 3848705 portforward.go:388] Datastream Reading Complete...Closing remoteDone channel
I0216 06:04:38.547230 3848705 portforward.go:420] handle FINISHED...
I0216 06:04:38.547241 3848705 websocket-connection.go:161] RemoveStream: 801
I0216 06:04:38.547600 3848705 websocket-connection.go:161] RemoveStream: 800
I0216 06:05:25.615538 3848705 portforward.go:395] Closing DataStream...Writing Complete
I0216 06:05:25.615867 3848705 portforward.go:388] Datastream Reading Complete...Closing remoteDone channel
I0216 06:05:25.615899 3848705 portforward.go:420] handle FINISHED...
I0216 06:05:25.615954 3848705 websocket-connection.go:161] RemoveStream: 1
I0216 06:05:25.615972 3848705 websocket-connection.go:161] RemoveStream: 0
```

The final removal of stream 0 and 1 is when the blocked endpoint released.

Reply:
> > are you saying you think they will block with this impl, but we don't want them to, and the spdy version would not?
> 
> If you have to manage fixed size buffers in the apiserver (connected to the websocket), and you have slow readers on the kubelet side (connected via SPDY), if a client then sends data to the apiserver and it makes it onto the websocket channel, the entire websocket is "blocked" until the buffer is emptied to the kubelet over SPDY.

The websocket connection does not block, since the websocket stream data is buffered ... up to a point. Very large port-forward requests could be problematic. When we ran a stress test of 1 million requests (and ~2 million streams) over several hours, the current implementation succeeded with zero errors.

![Screenshot from 2024-02-11 12-12-16](https://github.com/kubernetes/kubernetes/assets/7387162/ff924ae8-b4d8-4c0e-b05a-1b9a5f279221)

![Screenshot from 2024-02-11 12-12-42](https://github.com/kubernetes/kubernetes/assets/7387162/5f8b7a7f-97cf-4bf4-9aad-f0f9ec62be1c)

The request concurrency limit reached ~250 (~500 streams) within the apiserver during this test (as measured by the maximum size of the `requestStreamsMap`).

> 
> 1. Client 1 sends N bytes to server on port 1
> 2. Server reads N bytes off websocket to buffer
> 3. Server tries to send N bytes to kubelet
> 4. Kubelet accepts N bytes, SPDY window is reduced to 0 from apiserver
> 5. Kubelet tries to write N bytes to connection, connection is blocked
> 6. Client 1 sends N bytes to server
> 7. Server reads N bytes off websocket to buffer
> 8. Server tries to send N bytes to kubelet, but window is zero
> 9. Client 1 sends N bytes to server
> 10. Server reads N bytes off websocket, tries to write to buffer, buffer is full
> 11. Server is now blocked - it can't get rid of the buffer (without allocating more memory), it's can't drop the bytes (no client retry protocol)
> 12. Client 2 sends N bytes to server on port 2, but is blocked until remote client reads N bytes on port 1 and unblockes server

Reply:
> Very large requests may be problematic.

Is that websocket data buffered in the kernel, the goruntime, or our own code?  64kb isn't very large (that's what, 2 of the internal buffers, plus whatever happens in the network stack)?  In your test if we keep increasing the request size, at what point does it block (1MB?  10MB?  1GB?)

> For alpha, however, I decided to remove this request throttling because it wasn't clear it was necessary, and it would need significant benchmarks to get the concurrency limit correct. We can revisit this decision if needed.

Right, but for beta you have to have the throttling because otherwise this is a memory exhaustion DOS against the apiserver?

Reply:
> > Very large requests may be problematic.
> 
> Is that websocket data buffered in the kernel, the goruntime, or our own code? 640kb isn't very large (that's what, 2 of the internal buffers, plus whatever happens in the network stack)? In your test if we keep increasing the request size, at what point does it block (1MB? 10MB? 1GB?)
> 

On my dev machine using a local kind cluster, the buffering is filled somewhere between ~2.5MB (2621508) and ~2.6MB (2700860) bytes when SPDY is blocked.

### Full Methodology

#### Control

First, we ensured that the number of bytes being sent from the client were arriving at the test server. Using `curl` along with an endpoint at the test-server (`/countbytes`) which counted the size of the HTTP request body, we returned this body size in the response.
```
$ curl -v -X POST -d @data/post-2.5mb.dat localhost:3000/countbytes
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:3000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:3000...
* Connected to localhost (::1) port 3000
> POST /countbytes HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Length: 2621508
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
> 
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< Date: Mon, 19 Feb 2024 00:18:56 GMT
< Content-Length: 28
< Content-Type: text/plain; charset=utf-8
< 
num bytes received: 2621508
* Connection #0 to host localhost left intact
```
The number of bytes `curl` sent (2621508) was the same number arriving at the test server.

#### Blocking a 2.5MB request, other requests succeed
Hitting the `/blocking` endpoint sleeps for 60 seconds before returning the `blocked file` string.
```
$ curl -v -X POST -d @data/post-2.5mb.dat localhost:3000/blocking
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:3000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:3000...
* Connected to localhost (::1) port 3000
> POST /blocking HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Length: 2621508
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
> 
* Done waiting for 100-continue
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< Date: Sun, 18 Feb 2024 21:30:27 GMT
< Content-Length: 13
< Content-Type: text/plain; charset=utf-8
< Connection: close
< 
```
...waiting for the 60 second sleep. then
```
blocked file
* Closing connection
```
in another terminal running concurrently...
```
$ ./portforward-test-blocking 
Starting port forward test: 3000:3333 4000:3333

Temp Dir: /tmp/tmp.72aAHw9Lrk
Errors Dir: /tmp/tmp.3fzxBiezvp

.........10
.........20
.........30
.........40
.........50
.........60
.........70
.........80
.........90
.........100

Num Requests: 400
Num Errors: 0

Cleaning up...

Finished
```

#### Blocking a 2.6MB request, other requests are blocked
```
$ curl -v -X POST -d @data/post-2.6mb.dat localhost:3000/blocking
Note: Unnecessary use of -X or --request, POST is already inferred.
* Host localhost:3000 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:3000...
* Connected to localhost (::1) port 3000
> POST /blocking HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/8.5.0
> Accept: */*
> Content-Length: 2700860
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
> 
* Done waiting for 100-continue
* We are completely uploaded and fine
```
in another terminal running concurrently...
```
$ ./portforward-test-blocking 
Starting port forward test: 3000:3333 4000:3333

Temp Dir: /tmp/tmp.MPAtlUxmLN
Errors Dir: /tmp/tmp.ufF34guJWX

.^C   <BLOCKED & HUNG>
```

> > For alpha, however, I decided to remove this request throttling because it wasn't clear it was necessary, and it would need significant benchmarks to get the concurrency limit correct. We can revisit this decision if needed.
> 
> Right, but for beta you have to have the throttling because otherwise this is a memory exhaustion DOS against the apiserver?
> 
> edit: i saw it was 620kb - what point does it fail (please test locally to the breakpoint - if it never breaks, that's surprising to me).

Reply:
> > Very large requests may be problematic.
> 
> Is that websocket data buffered in the kernel, the goruntime, or our own code? 640kb isn't very large (that's what, 2 of the internal buffers, plus whatever happens in the network stack)? In your test if we keep increasing the request size, at what point does it block (1MB? 10MB? 1GB?)
> 
> > For alpha, however, I decided to remove this request throttling because it wasn't clear it was necessary, and it would need significant benchmarks to get the concurrency limit correct. We can revisit this decision if needed.
> 
> Right, but for beta you have to have the throttling because otherwise this is a memory exhaustion DOS against the apiserver?
> 

Yes. We plan to address this issue in Beta. In our KEP, we labeled this risk [*Overloaded Concurrency*](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets#risks-and-mitigations). We need to benchmark and test a request concurrency limit which would protect the API Server, but not unduly diminish the performance of current SPDY port-forwards which work today. By submitting the current implementation as Alpha, we can allow a broad array of opt-in testers to get good data on this issue.

Please let me know what you think of this approach.

> edit: i saw it was 620kb - what point does it fail (please test locally to the breakpoint - if it never breaks, that's surprising to me).